fuel-relayer/src/relayer/get_logs.rs

```diff
@@ -0,0 +1,72 @@
+use super::*;
+use futures::TryStreamExt;
+
+#[cfg(test)]
+mod test;
+
+/// Download the logs from the DA layer.
+pub(crate) fn download_logs<P>(
+    eth_sync_gap: &state::EthSyncGap,
+    contracts: Vec<H160>,
+    eth_node: Arc<P>,
+    page_size: u64,
+) -> impl futures::Stream<Item = Result<Vec<Log>, ProviderError>>
+where
+    P: Middleware<Error = ProviderError> + 'static,
+{
+    // Create a stream of paginated logs.
+    futures::stream::try_unfold(
+        eth_sync_gap.page(page_size),
+        move |mut page: state::EthSyncPage| {
+            let contracts = contracts.clone();
+            let eth_node = eth_node.clone();
+            async move {
+                if page.is_empty() {
+                    Ok(None)
+                } else {
+                    // Create the log filter from the page.
+                    let filter = Filter::new()
+                        .from_block(page.oldest())
+                        .to_block(page.latest())
+                        .address(ValueOrArray::Array(contracts));
+
+                    // Reduce the page.
+                    page.reduce();
+
+                    // Get the logs and return the reduced page.
+                    eth_node
+                        .get_logs(&filter)
+                        .await
+                        .map(|logs| Some((logs, page)))
+                }
+            }
+        },
+    )
+}
+
+/// Write the logs to the database.
+pub(crate) async fn write_logs<S>(
+    database: &mut dyn RelayerDb,
+    logs: S,
+) -> anyhow::Result<()>
+where
+    S: futures::Stream<Item = Result<Vec<Log>, ProviderError>>,
+{
+    tokio::pin!(logs);
+    while let Some(events) = logs.try_next().await? {
```

Comment:
Yeh I actually intentionally did this because I wasn't sure about adding the complexity on tracking pagination. Maybe this would suite a follow up PR as it would be nice to get it reviewed properly. I did assume that downloading messages was idempotent but it sounds like that's not the case.

Reply:
I'm fine with deferring this to a follow-up issue, feel free to open new task(s) for these changes to resolve this thread if you'd prefer to address this later.

Reply:
TODO: Make issue for the second concern.

Reply:
I've addressed the first concern.